File: sdks/node/test/semantic-dependency.test.js
import assert from "node:assert/strict";
import test from "node:test";

import {
  createManagedEngineProjectForTest,
  runOperation,
} from "../src/engine-operation.js";
import {
  dependencyRequest,
  dependencyResponse,
  runDependency,
} from "../src/semantic-dependency.js";

function request(changes = {}) {
  return dependencyRequest({
    encoding: "http-1.1-message",
    metadata: [
      { name: "x-tag", value: Buffer.from("first") },
      { name: "x-tag", value: Buffer.from("second") },
    ],
    method: "POST",
    observationClass: "outbound-http",
    operation: "outbound-http-request",
    payload: Buffer.from("request"),
    protocol: "http-1.1",
    target: "https://inventory.example/item",
    ...changes,
  });
}

function response(changes = {}) {
  return dependencyResponse({
    metadata: [
      { name: "x-tag", value: Buffer.from("first") },
      { name: "x-tag", value: Buffer.from("second") },
    ],
    outcome: "response",
    payload: Buffer.from("response"),
    statusCode: 200,
    ...changes,
  });
}

function responseRecord(value) {
  return Buffer.from(JSON.stringify({
    error_code: value.errorCode,
    error_number: value.errorNumber,
    metadata: value.metadata.map((field) => ({
      name: Buffer.from(field.name).toString("base64url"),
      value: field.value.toString("base64url"),
    })),
    outcome: value.outcome,
    payload: value.payload === null ? null : value.payload.toString("base64url"),
    status: value.status,
    status_code: value.statusCode,
  }));
}

class DependencyBridge {
  calls = [];
  failFinish = false;
  failOpen = false;
  failRead = false;
  #action;
  #offset = 0;
  #readBytes;
  #response;

  constructor(action, responseBytes = Buffer.alloc(0), readBytes = 17) {
    this.#action = action;
    this.#readBytes = readBytes;
    this.#response = responseBytes;
  }

  operationBegin() {
    return { operationHandle: 2, operationId: "op_dependency" };
  }

  operationInput() {}

  dependencyOpen(handle, dependencyRequestValue, parent) {
    if (this.failOpen) throw new Error("private bridge failure");
    this.calls.push(["dependency-open", handle, dependencyRequestValue, parent]);
    return { action: this.#action, dependencyHandle: 4 };
  }

  observationRead(handle) {
    if (this.failRead) throw new Error("private bridge failure");
    const end = Math.min(this.#offset + this.#readBytes, this.#response.length);
    const chunk = this.#response.subarray(this.#offset, end);
    this.#offset = end;
    this.calls.push(["observation-read", handle, chunk.length]);
    return { chunk, eof: this.#offset === this.#response.length };
  }

  dependencyFinish(handle, dependencyResponseValue) {
    if (this.failFinish) throw new Error("private bridge failure");
    this.calls.push(["dependency-finish", handle, dependencyResponseValue]);
    return dependencyResponseValue?.outcome ?? "response";
  }

  operationSucceed() {}

  operationAbandon(handle) {
    this.calls.push(["operation-abandon", handle]);
  }

  engineClose() {}
}

test("capture uses the exact native call shape for response and error", () => {
  for (const captured of [
    response(),
    response({
      errorCode: "not-found",
      errorNumber: 2,
      metadata: [],
      outcome: "error",
      payload: null,
      statusCode: null,
    }),
  ]) {
    const bridge = new DependencyBridge("capture");
    assert.equal(run(bridge, () => captured), captured);
    const [open, finish] = bridge.calls;
    assert.deepEqual(open.slice(0, 2), ["dependency-open", 2]);
    assert.equal(open[2].payload, "cmVxdWVzdA");
    assert.deepEqual(open[2].metadata, [
      { name: "eC10YWc", value: "Zmlyc3Q" },
      { name: "eC10YWc", value: "c2Vjb25k" },
    ]);
    assert.deepEqual(finish.slice(0, 2), ["dependency-finish", 4]);
    assert.equal(finish[2].outcome, captured.outcome);
  }
});

test("replay reads chunks, finishes validation, then reconstructs", () => {
  const expected = response();
  const bridge = new DependencyBridge("replay", responseRecord(expected), 7);
  let calls = 0;
  const actual = run(bridge, () => {
    calls += 1;
    return expected;
  });
  assert.deepEqual(actual, expected);
  assert.equal(calls, 0);
  assert.deepEqual(bridge.calls.at(-1), ["dependency-finish", 4, null]);
});

test("capture failures preserve one exact result or exception", () => {
  for (const mode of ["conversion", "open", "finish"]) {
    const selected = mode === "conversion"
      ? request({ payload: Buffer.alloc(65_537) })
      : request();
    const resultBridge = new DependencyBridge("capture");
    resultBridge.failOpen = mode === "open";
    resultBridge.failFinish = mode === "finish";
    const expected = response();
    let resultCalls = 0;
    const actual = run(resultBridge, () => {
      resultCalls += 1;
      return expected;
    }, selected);
    assert.equal(actual, expected, `${mode} result identity`);
    assert.equal(resultCalls, 1, `${mode} result calls`);

    const errorBridge = new DependencyBridge("capture");
    errorBridge.failOpen = mode === "open";
    errorBridge.failFinish = mode === "finish";
    const sentinel = new Error("application sentinel");
    let errorCalls = 0;
    assert.throws(() => run(errorBridge, () => {
      errorCalls += 1;
      throw sentinel;
    }, selected), (error) => error === sentinel);
    assert.equal(errorCalls, 1, `${mode} error calls`);
  }
});

test("async capture preserves the exact result", async () => {
  const bridge = new DependencyBridge("capture");
  const expected = response();
  assert.equal(await run(bridge, async () => expected), expected);
});

test("metadata conversion is bounded before engine open", () => {
  const bridge = new DependencyBridge("capture");
  const expected = response();
  const selected = request({
    metadata: [{ name: "name", value: Buffer.alloc(65_537) }],
  });
  assert.equal(run(bridge, () => expected, selected), expected);
  assert.equal(bridge.calls.some((call) => call[0] === "dependency-open"), false);
});

test("replay failures never call the live dependency", () => {
  for (const mode of ["read", "finish", "record"]) {
    const bridge = new DependencyBridge("replay", Buffer.from("not-json"));
    bridge.failRead = mode === "read";
    bridge.failFinish = mode === "finish";
    let calls = 0;
    assert.throws(() => run(bridge, () => {
      calls += 1;
      return response();
    }), { code: "ERR_REPROIT_SEMANTIC_DEPENDENCY" });
    assert.equal(calls, 0);
  }
});

function run(bridge, live, selected = request()) {
  const project = createManagedEngineProjectForTest(bridge, 1, () => "unused");
  return runOperation(
    project,
    { begin: {}, completion: "return", inputs: [] },
    () => runDependency(selected, live),
    () => null,
  );
}

File: sdks/go/reproit/semantic_dependency_test.go
package reproit

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"errors"
	"os"
	"reflect"
	"testing"
)

func TestSDKEngineDependencyBridgeUsesExactCallShape(t *testing.T) {
	var calls []map[string]any
	bridge := operationTestBridge(t, func(input []byte, output []byte) int64 {
		var call map[string]any
		_ = json.Unmarshal(input, &call)
		calls = append(calls, call)
		result := `{}`
		if call["operation"] == sdkEngineDependencyOpen {
			result = `{"action":"capture","dependency_handle":41}`
		} else if call["operation"] == sdkEngineDependencyFinish {
			result = `{"outcome":"response"}`
		}
		return int64(copy(output, sdkEngineSuccess(result)))
	})
	defer bridge.close()
	parent := "op_parent"
	started, err := bridge.openDependency(7, &parent, sdkEngineDependencyRequest{
		Encoding: "bytes",
		Metadata: []sdkEngineDependencyMetadata{
			{Name: "eC10YWc", Value: "Zmlyc3Q"},
			{Name: "eC10YWc", Value: "c2Vjb25k"},
		},
		ObservationClass: "outbound-http",
		Operation:        "outbound-http-request",
		Payload:          "cGF5bG9hZA",
		Protocol:         "http-1.1",
		Target:           "aHR0cHM6Ly9leGFtcGxlLmNvbQ",
	})
	if err != nil || started.Action != "capture" || started.Handle != 41 {
		t.Fatal("The dependency-open result was rejected.")
	}
	payload := ""
	outcome, err := bridge.finishDependency(started.Handle, &sdkEngineDependencyResponse{
		Metadata: []sdkEngineDependencyMetadata{}, Outcome: "response", Payload: &payload,
	})
	if err != nil || outcome != "response" {
		t.Fatal("The dependency-finish result was rejected.")
	}
	open := calls[0]
	request := open["request"].(map[string]any)
	metadata := request["metadata"].([]any)
	if !hasExactKeys(open,
		"causal_parent_id", "format", "operation", "operation_handle", "request") ||
		!hasExactKeys(request,
			"encoding", "metadata", "method", "observation_class", "operation", "payload",
			"protocol", "target") || open["causal_parent_id"] != parent ||
		metadata[0].(map[string]any)["value"] != "Zmlyc3Q" ||
		metadata[1].(map[string]any)["value"] != "c2Vjb25k" {
		t.Fatal("The dependency bridge changed its exact call shape or metadata order.")
	}
	finish := calls[1]
	if !hasExactKeys(finish, "dependency_handle", "format", "operation", "response") {
		t.Fatal("The dependency-finish call has unexpected fields.")
	}
}

func TestSemanticDependencyCapturePreservesLiveSuccessAndError(t *testing.T) {
	for _, outcome := range []observationOutcome{observationResponse, observationError} {
		t.Run(string(outcome), func(t *testing.T) {
			var calls []string
			bridge := semanticDependencyBridge(t, "capture", nil, func(call map[string]any) string {
				name := call["operation"].(string)
				calls = append(calls, name)
				if name == sdkEngineDependencyFinish {
					return sdkEngineSuccess(`{"outcome":"` + string(outcome) + `"}`)
				}
				return ""
			})
			defer bridge.close()
			response := semanticTestResponse(outcome)
			sentinel := errors.New("sentinel dependency error")
			var liveErr error
			if outcome == observationError {
				liveErr = sentinel
			}
			liveCalls := 0
			got, gotErr := translateSemanticDependency(
				semanticTestOperation(bridge), semanticTestRequest(), nil,
				func() (semanticDependencyResponse, error) {
					liveCalls++
					return response, liveErr
				},
			)
			if liveCalls != 1 || gotErr != liveErr || !reflect.DeepEqual(got, response) ||
				!reflect.DeepEqual(calls, []string{sdkEngineDependencyOpen, sdkEngineDependencyFinish}) {
				t.Fatal("Capture changed the exact live result, error, or engine sequence.")
			}
		})
	}
}

func TestSemanticDependencyCaptureFailsOpenAtEveryEngineBoundary(t *testing.T) {
	tests := []struct {
		name             string
		failOperation    string
		oversizedRequest bool
		oversizedResult  bool
	}{
		{name: "request conversion", oversizedRequest: true},
		{name: "dependency open", failOperation: sdkEngineDependencyOpen},
		{name: "response conversion", oversizedResult: true},
		{name: "dependency finish", failOperation: sdkEngineDependencyFinish},
	}
	for _, test := range tests {
		for _, withError := range []bool{false, true} {
			t.Run(test.name, func(t *testing.T) {
				bridge := semanticDependencyBridge(t, "capture", nil, func(call map[string]any) string {
					if call["operation"] == test.failOperation {
						return semanticEngineRejected()
					}
					return ""
				})
				defer bridge.close()
				request := semanticTestRequest()
				if test.oversizedRequest {
					request.Metadata = []semanticDependencyMetadata{{
						Name: make([]byte, sdkEngineMaxCallBytes+1),
					}}
				}
				response := semanticTestResponse(observationResponse)
				if test.oversizedResult {
					response.Metadata = []semanticDependencyMetadata{{
						Name: make([]byte, sdkEngineMaxCallBytes+1),
					}}
				}
				sentinel := errors.New("sentinel dependency error")
				var wantErr error
				if withError {
					wantErr = sentinel
				}
				liveCalls := 0
				got, gotErr := translateSemanticDependency(
					semanticTestOperation(bridge), request, nil,
					func() (semanticDependencyResponse, error) {
						liveCalls++
						return response, wantErr
					},
				)
				if liveCalls != 1 || gotErr != wantErr || !reflect.DeepEqual(got, response) {
					t.Fatal("A capture failure changed the exact live result or error.")
				}
			})
		}
	}
}

func TestSemanticDependencyReplayReadsChunksAfterEngineValidation(t *testing.T) {
	record := semanticPublishedResponse(t, "semantic_dependency_response_outbound_http")
	split := len(record) / 2
	reads := [][]byte{record[:split], record[split:]}
	bridge := semanticDependencyBridge(t, "replay", reads, nil)
	defer bridge.close()
	liveCalled := false
	response, err := translateSemanticDependency(
		semanticTestOperation(bridge), semanticTestRequest(), nil,
		func() (semanticDependencyResponse, error) {
			liveCalled = true
			return semanticDependencyResponse{}, nil
		},
	)
	if err != nil || liveCalled || response.Outcome != observationResponse ||
		response.StatusCode == nil || *response.StatusCode != 200 ||
		!response.HasPayload || string(response.Payload) != `{"available":true}` {
		t.Fatal("Replay did not reconstruct the engine-validated dependency response.")
	}
}

func TestSemanticDependencyUsesEngineValidatedOutcome(t *testing.T) {
	record := semanticPublishedResponse(t, "semantic_dependency_response_outbound_http")
	var wire map[string]any
	if json.Unmarshal(record, &wire) != nil {
		t.Fatal("The published dependency response could not be decoded.")
	}
	wire["outcome"] = "error"
	record, _ = json.Marshal(wire)
	response, err := reconstructSemanticDependencyResponse(record, "response")
	if err != nil || response.Outcome != observationResponse {
		t.Fatal("The language bridge duplicated the engine outcome validation.")
	}
}

func TestSemanticDependencyReplayNeverFallsBackToLive(t *testing.T) {
	tests := []struct {
		name          string
		reads         [][]byte
		finishFailure bool
	}{
		{name: "empty read", reads: [][]byte{{}}},
		{name: "record over bound", reads: [][]byte{
			make([]byte, sdkEngineMaxSemanticDependencyRecordBytes), {1},
		}},
		{name: "finish rejection", reads: [][]byte{[]byte(`{}`)}, finishFailure: true},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			bridge := semanticDependencyBridge(t, "replay", test.reads, func(call map[string]any) string {
				if test.finishFailure && call["operation"] == sdkEngineDependencyFinish {
					return semanticEngineRejected()
				}
				return ""
			})
			defer bridge.close()
			liveCalled := false
			_, err := translateSemanticDependency(
				semanticTestOperation(bridge), semanticTestRequest(), nil,
				func() (semanticDependencyResponse, error) {
					liveCalled = true
					return semanticDependencyResponse{}, nil
				},
			)
			if !errors.Is(err, ErrAutomaticCapture) || liveCalled {
				t.Fatal("Strict replay fell back to the live dependency.")
			}
		})
	}
}

func semanticDependencyBridge(
	t *testing.T,
	action string,
	reads [][]byte,
	intercept func(map[string]any) string,
) *sdkEngineBridge {
	t.Helper()
	readIndex := 0
	return operationTestBridge(t, func(input []byte, output []byte) int64 {
		var call map[string]any
		_ = json.Unmarshal(input, &call)
		if intercept != nil {
			if result := intercept(call); result != "" {
				return int64(copy(output, result))
			}
		}
		result := `{}`
		switch call["operation"] {
		case sdkEngineDependencyOpen:
			result = `{"action":"` + action + `","dependency_handle":71}`
		case sdkEngineObservationRead:
			chunk := reads[readIndex]
			readIndex++
			result = `{"chunk":"` + base64.RawURLEncoding.EncodeToString(chunk) +
				`","eof":` + map[bool]string{true: "true", false: "false"}[readIndex == len(reads)] + `}`
		case sdkEngineDependencyFinish:
			result = `{"outcome":"response"}`
		}
		return int64(copy(output, sdkEngineSuccess(result)))
	})
}

func semanticTestRequest() semanticDependencyRequest {
	method := "POST"
	return semanticDependencyRequest{
		Encoding: "http-1.1-message",
		Metadata: []semanticDependencyMetadata{
			{Name: []byte("x-tag"), Value: []byte("first")},
			{Name: []byte("x-tag"), Value: []byte("second")},
		},
		Method:           &method,
		ObservationClass: observationOutboundHTTP,
		Operation:        "outbound-http-request",
		Payload:          []byte(`{"quantity":1}`),
		Protocol:         "http-1.1",
		Target:           "https://inventory.example/v1/items/42",
	}
}

func semanticTestResponse(outcome observationOutcome) semanticDependencyResponse {
	if outcome == observationError {
		code := "other"
		return semanticDependencyResponse{ErrorCode: &code, Outcome: outcome}
	}
	status := uint16(200)
	return semanticDependencyResponse{
		Metadata: []semanticDependencyMetadata{{
			Name: []byte("content-type"), Value: []byte("application/json"),
		}},
		Outcome:    outcome,
		Payload:    []byte(`{"available":true}`),
		HasPayload: true,
		StatusCode: &status,
	}
}

func semanticPublishedResponse(t *testing.T, name string) []byte {
	t.Helper()
	encoded, err := os.ReadFile(os.Getenv("REPROIT_PROTOCOL_VECTORS"))
	if err != nil {
		t.Fatal("The published protocol vectors are unavailable.")
	}
	decoder := json.NewDecoder(bytes.NewReader(encoded))
	decoder.UseNumber()
	var vectors struct {
		Positive map[string]struct {
			Value map[string]any `json:"value"`
		} `json:"positive"`
	}
	if decoder.Decode(&vectors) != nil {
		t.Fatal("The published protocol vectors are invalid.")
	}
	record, err := CanonicalBytes(vectors.Positive[name].Value)
	if err != nil {
		t.Fatal(err)
	}
	return record
}

func semanticTestOperation(bridge *sdkEngineBridge) *AutomaticOperation {
	project := &AutomaticProject{bridge: bridge}
	return &AutomaticOperation{handle: 1, operationID: "op_semantic", project: project}
}

func semanticEngineRejected() string {
	return `{"error_code":"SCHEMA_INVALID","format":"` + engineResponseFormat +
		`","ok":false,"result":{}}`
}
